### PR TITLE
Support for DRUPAL_SETTINGS as an override for Drupal configuration

### PR DESCRIPTION
**Pull request by populist**
There are many different Drupal development and hosting platforms that would like to inject their settings into a Drupal installation without formally writing it to the settings.php.

This pull request allows this through a $_SERVER['DRUPAL_SETTINGS'] value that can be injected in a variety of ways. This is the method we have used at Pantheon to inject our settings for years, but would love for it to be more cleanly supported in Backdrop.

### CODE REVIEW

**Comment by fluxsauce** [comment]
:+1: for me. Full disclosure, I work at Pantheon, but this is pretty much the same as https://github.com/pressflow/7/blob/master/includes/bootstrap.inc#L720 and it's very tried and true.

Now, if there's a concern about the use of DRUPAL_SETTINGS, something like SERVER_SETTINGS or ENV_CONF could also work.

**Comment by quicksketch** [comment]
Hi guys, thanks for the PR. A few requests:
- Let's avoid "Drupal" in the variable name. We're not sure what we're going to do about the word Drupal yet, but I'd prefer to avoid introducing new situations. (https://github.com/backdrop/backdrop-issues/issues/3)
- Please make an associated issue for tracking this, in the event that this PR needs to be replaced with a different one, and so we can tag and track it.
- We definitely need to introduce tests for checking functionality that almost by nature will have very little documentation or end-user facing functionality.

**Comment by populist** [comment]
Done -

1 - DRUPAL_SETTINGS has been changed to SERVER_SETTINGS
2 -  Associated issue - https://github.com/backdrop/backdrop-issues/issues/90

Still To Do - 

3 - Tests

File: core/includes/bootstrap.inc [hunk 1]
```diff
@@ -771,6 +771,30 @@ function drupal_settings_initialize() {
   }
   $is_https = isset($_SERVER['HTTPS']) && strtolower($_SERVER['HTTPS']) == 'on';
 
+  // Load Drupal settings specified by the server, if present
```
Comment (sirkitree) [inline core/includes/bootstrap.inc:774]: Minor point, but should have a period at the end of the sentence here.

**Comment by sirkitree** [comment]
Also wondering about documentation. Where is the best place to document how this is used in a visible way so people know it's available or can be extracted by automatic documentation tools?

**Comment by quicksketch** [comment]
I've been thinking about how to test this functionality and I think I came up with a way for this to work.

We already pull in a custom php.ini and vhost.conf file for TravisCI to do test. Let's just put in custom configuration data in those files and leave them there all the time. Then we can make a test that checks for the existence of those settings. If the check flag is entirely missing from the server configuration, we can skip the test entirely. This would allow us to test server-provided database configuration, conf settings, and any other global variables that are in settings.php without much trouble and with no significant changes to our testing.

Documentation-wise, I think putting a small amount of docs in `drupal_settings_initialize` should be adequate, or we should link to online documentation about the feature. Unfortunately we don't have an API website where such documentation would live, so for now let's do what we can in the docblock of the containing function.

**Comment by jenlampton** [comment]
@populist did you see the suggestion for how to test? 

> We already pull in a custom php.ini and vhost.conf file for TravisCI to do test. Let's just put in
> custom configuration data in those files and leave them there all the time. Then we can make a 
> test that checks for the existence of those settings. If the check flag is entirely missing from the 
> server configuration, we can skip the test entirely. This would allow us to test server-provided 
> database configuration, conf settings, and any other global variables that are in settings.php 
> without much trouble and with no significant changes to our testing.

**Comment by gf-pantheon** [comment]
@jenlampton I saw the testing information and think that could work. Will require a bit of coding, but seems reasonable enough.

In the other world, check out https://drupal.org/node/1830816 for some sense of how Drupal 7/8 is tackling the problem which seems similar enough aside from the array key.

**Comment by quicksketch** [comment]
This is seeming like an increasingly common approach to server-specific configuration. I'm no expert at Heroku apps (just getting started on another project), but it looks like they use the same approach for setting environment-specific variables. The fact that they don't use PHP (they're Ruby and node.js for the most part) isn't particularly relevant, but the exact same approach is used to pull out values to use them in an application. The most common usage for this happens to be the database connection string (in my limited experience). More info at https://devcenter.heroku.com/articles/how-heroku-works#config-vars

This really should be made it's own issue so we don't have to keep scrounging around in the PR queue for this request, and this PR might need to be replaced with a new one.

**Comment by quicksketch** [comment]
I'm working on integrating this functionality, but it's dependent on backdrop/backdrop-issues#280, in which we're adding a new $settings variable. Overriding $conf doesn't make any sense because it's not going to exist by the time the final version of Backdrop comes out.

**Comment by quicksketch** [comment]
I made new issue to track this, since I need to file a new PR that includes tests. Let's continue tracking in https://github.com/backdrop/backdrop-issues/issues/281.

**Pull request closed without merging**